lib/rlpx/peer.js

```diff
@@ -0,0 +1,285 @@
+  let payload = rlp.decode(msg)
+  switch (code) {
+    case PREFIXES.HELLO:
+      this._hello = {
+        protocolVersion: util.buffer2int(payload[0]),
+        clientId: payload[1].toString(),
+        capabilities: payload[2].map((item) => {
+          return { name: item[0].toString(), version: util.buffer2int(item[1]) }
+        }),
+        port: util.buffer2int(payload[3]), // Q: always should be 16-bits ?
+        id: payload[4]
+      }
+      this._remoteId = new Buffer(this._hello.id)
+
+      let shared = {}
+      for (let item of this._hello.capabilities) {
+        for (let obj of this._capabilities) {
+          if (obj.name !== item.name || obj.version !== item.version) continue
+          if (shared[obj.name] && shared[obj.name].version > obj.version) continue
+          shared[obj.name] = obj
+        }
+      }
+
+      let offset = BASE_PROTOCOL_LENGTH
+      this._protocols = Object.keys(shared).map((key) => shared[key])
+        .sort((obj1, obj2) => obj1.name < obj2.name ? -1 : 1)
+        .map((obj) => {
+          let _offset = offset
+          offset += obj.length
+
+          let SubProtocol = obj.constructor
+          let protocol = new SubProtocol(obj.version, this, (code, data) => {
+            if (code > obj.length) throw new Error('Code out of range')
+            this._sendMessage(Buffer.concat([ new Buffer([ _offset + code ]), data ]))
+          })
+
+          return { protocol, offset: _offset, length: obj.length }
+        })
+
+      if (this._protocols.length === 0) {
+        return this._sendDisconnect(Peer.DISCONNECT_REASONS.INCOMPATIBLE_P2P_PROTOCOL_VERSION)
+      }
+
+      this._connected = true
+      this._pingIntervalId = setInterval(() => this._sendPing(), PING_INTERVAL)
+      this.emit('connect')
+      break
+
+    case PREFIXES.DISCONNECT:
+      this._closed = true
+      this._disconnectReason = payload[0][0]
+      this._socket.end()
+      break
+
+    case PREFIXES.PING:
+      this._sendPong()
+      break
+
+    case PREFIXES.PONG:
+      clearTimeout(this._pingTimeoutId)
+      break
+  }
+}
+
+Peer.prototype._sendAuth = function () {
+  if (this._closed) return
+  this._socket.write(this._eciesSession.createAuth())
+  this._state = 'Ack'
+  this._nextPacketSize = 210
+}
+
+Peer.prototype._sendAck = function () {
+  if (this._closed) return
+  this._socket.write(this._eciesSession.createAck())
+  this._state = 'Header'
+  this._nextPacketSize = 32
+  this._sendHello()
+}
+
+Peer.prototype._sendMessage = function (msg) {
+  if (this._closed) return false
+  this._socket.write(this._eciesSession.createHeader(msg.length))
+  this._socket.write(this._eciesSession.createBody(msg))
+  return true
+}
+
+Peer.prototype._sendHello = function () {
+  let data = [
+    util.int2buffer(BASE_PROTOCOL_VERSION),
+    this._clientId,
+    this._capabilities.map((obj) => [ new Buffer(obj.name), util.int2buffer(obj.version) ]),
+    util.int2buffer(this._port), // Q: always should be 16-bits ?
```

Comment:
rlp I think. But there isn't a clear spec. So I you may just have to look at go-ethereum code to find out for sure.